pytorch_lightning/accelerators/gpu.py

```diff
@@ -37,8 +37,14 @@ def setup_environment(self, root_device: torch.device) -> None:
         """
         Raises:
             MisconfigurationException:
+                If torch.cuda isn't available.
+                If no CUDA devices are found.
                 If the selected device is not GPU.
         """
+        if not torch.cuda.is_available():
+            raise MisconfigurationException("GPU Accelerator used, but CUDA isn't available.")
+        if torch.cuda.device_count() == 0:
+            raise MisconfigurationException("GPU Accelerator used, but found no CUDA devices available.")
```

Comment:
quick ques: why not inside accelerator init?

Reply:
User should be able to create a accelerator object in any environment, we only want to check device availability when call trainer.fit/test/eval, and setup_evn is the first hook called in trainer._run

Reply:
I would also prefer the `__init__`, it's generally where strategies run their checks:

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/strategies/ipu.py#L98-L102

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/strategies/deepspeed.py#L263-L266

Reply:
> User should be able to create a accelerator object in any environment, we only want to check device availability when call trainer.fit/test/eval, and setup_evn is the first hook called in trainer._run

quick ques: can you explain more on this? do you mean creating obj in non-GPU env and using it in GPU env? is that possible?

Reply:
@rohitgr7 I think it's possible in production/complex system, for example: User needs to create obj in a config script, and then send the config file to a scheduler to trigger/schedule the training job.

Reply:
>But to the best of my knowledge, the config files define the structure of the object constructor but don't actually initialize them until the job is executed.

Agree. :) 
I saw our internal users/team have config wrappers on top of lightning (for reusability), where defined accelerator obj. I'm not sure whether it will fail if we add check in `__init__`, but feel better to be safe.

Reply:
I can't recall any instance where the GPU accelerator is specified on a host which does not have GPU devices. In almost every instance I've seen, the entire training script is packaged and run on a GPU host.

I moved the assertion to happen back during `__init__` to fail faster